SpatialGDK/Source/SpatialGDK/Private/Interop/RPCs/CrossServerRPCService.cpp

```diff
@@ -0,0 +1,476 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "Interop/RPCs/CrossServerRPCService.h"
+
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "SpatialConstants.h"
+#include "Utils/RepLayoutUtils.h"
+
+DEFINE_LOG_CATEGORY(LogCrossServerRPCService);
+
+namespace SpatialGDK
+{
+CrossServerRPCService::CrossServerRPCService(const ActorCanExtractRPCDelegate InCanExtractRPCDelegate,
+											 const ExtractRPCDelegate InExtractRPCCallback, const FSubView& InSubView,
+											 FRPCStore& InRPCStore)
+	: CanExtractRPCDelegate(InCanExtractRPCDelegate)
+	, ExtractRPCCallback(InExtractRPCCallback)
+	, SubView(&InSubView)
+	, RPCStore(&InRPCStore)
+{
+}
+
+EPushRPCResult CrossServerRPCService::PushCrossServerRPC(Worker_EntityId EntityId, const RPCSender& Sender,
+														 const PendingRPCPayload& Payload, bool bCreatedEntity)
+{
+	CrossServerEndpoints* Endpoints = CrossServerDataStore.Find(Sender.Entity);
+	Schema_Object* EndpointObject = nullptr;
+	EntityComponentId SenderEndpointId(Sender.Entity, SpatialConstants::CROSSSERVER_SENDER_ENDPOINT_COMPONENT_ID);
+
+	if (!Endpoints)
+	{
+		if (bCreatedEntity)
+		{
+			return EPushRPCResult::EntityBeingCreated;
+		}
+
+		EndpointObject = Schema_GetComponentDataFields(RPCStore->GetOrCreateComponentData(SenderEndpointId));
+		Endpoints = &CrossServerDataStore.Add(Sender.Entity);
+	}
+	else
+	{
+		EndpointObject = Schema_GetComponentUpdateFields(RPCStore->GetOrCreateComponentUpdate(SenderEndpointId));
+	}
+
+	CrossServer::WriterState& SenderState = Endpoints->SenderState;
+
+	TOptional<uint32> Slot = SenderState.Alloc.PeekFreeSlot();
```

Comment:
Is it worth just attempting to reserve the slot here? To avoid potential API misuse by using a peeked slot without committing it. Perhaps it could safeguarded a little by ensuring that Reserve is followed by Commit and not Reserve again (in the case of an early out or something)

Reply:
True, guess the code changed as to make it not necessary to separate them.